pulp_rpm/app/migrations/0052_modulemd_digest.py

```diff
@@ -0,0 +1,51 @@
+# Generated by Django 4.2.4 on 2023-09-04 14:06
+
+from django.db import migrations, models
+import hashlib
+
+
+def add_snippet_hash(apps, schema_editor):
+    """Calculate and add digest hash of the snippet."""
+
+    Modulemd = apps.get_model("rpm", "Modulemd")
+    modules_to_update = []
+
+    for mmd in Modulemd.objects.all().only("snippet").iterator():
+        digest = hashlib.sha256(mmd.snippet.encode()).hexdigest()
+        modules_to_update.append(mmd)
+    Modulemd.objects.bulk_update(modules_to_update, fields=["digest"])
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ("rpm", "0051_alter_distributiontree_unique_together_and_more"),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name="modulemd",
+            name="digest",
+            field=models.TextField(default=""),
+        ),
+        migrations.RunPython(
+            code=add_snippet_hash,
+            reverse_code=migrations.RunPython.noop,
+            elidable=True,
+        ),
+        migrations.AlterUniqueTogether(
+            name="modulemd",
+            unique_together=set(),
+        ),
+        migrations.AlterField(
+            model_name="modulemd",
+            name="digest",
+            field=models.TextField(),
+        ),
+        migrations.AlterUniqueTogether(
+            name="modulemd",
+            unique_together={
+                ("_pulp_domain", "name", "stream", "version", "context", "arch", "digest")
```

Comment:
Should we set this to just ```("_pulp_domain", "digest")```?

Seems like the other fields wouldn't help much.

However we probably should keep indexes for name and stream.

Reply:
we could but that sort of breaks my plan how to mitigate modules with empty snippets?

Reply:
I assume by A you mean telling the user which repos to resync?  Or do you mean doing it in the migration?

Reply:
I meant doing it in the migration which is not my preference. In order to run this migration successfully the digest must be calculated. If there will be empty snippets the digest for those modules will be `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855` and migration will fail with:
```
django.db.utils.IntegrityError: could not create unique index "rpm_modulemd__pulp_domain_id_digest_eb1e206a_uniq"
DETAIL:  Key (_pulp_domain_id, digest)=(018a6a12-6b49-797c-9f7b-32c3da09e5c0, e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855) is duplicated.
```

Reply:
and B will not solve the problem as long as these kind of modules will stay in DB

Reply:
I don't think we can do it in the migration.  It may not even be reliable - what happens if the repo is gone, or the module isn't in the metadata anymore?  It introduces a lot of variables that shouldn't be anywhere near the migration process.

Reply:
I think my personal preference is to let the migration fail, but spit out the details of which repositories and modules are impacted and how it can be fixed [0].  That will make it easy for users to work around the issue while not limiting us unnecessarily (since we don't know how big of an issue this will ultimately end up being).

I don't know that it's the perfect solution, but it feels like a decent compromise between overengineering and being inconsiderate to users.  

[0] For instance, if it *is* present in the upstream repo then a resync could fix it.  If it's not, then a mirror-sync combined with orphan cleanup could fix it.  And the nuclear option is to just delete the repo, orphan cleanup, and resync, but probably one of the former two would work.

Reply:
>I don't think we can do it in the migration. 

I agree with you on that, but listed this as on option anyway.

>For instance, if it is present in the upstream repo then a resync could fix it. If it's not, then a mirror-sync combined with orphan cleanup could fix it. And the nuclear option is to just delete the repo, orphan cleanup, and resync, but probably one of the former two would work.

Just the repo removal,  orphan clean up, resync would work.
Given the in place uniqueness constant of  NSVCA on re-sync it would identify the module as an existing content unit and skip it.